### DISCUSSION TITLE
Why is context preferred over a store hook for dependency injection?

### DISCUSSION THREAD

**Discussion by jorgezreik** [pmndrs/zustand#1774] [Q&A] [answered]
The Zustand [dependency injection documentation](https://docs.pmnd.rs/zustand/guides/initialize-state-with-props) recommends using "a vanilla store with React.context" for dependency injection.

Why is this approach preferred over the following-hook based alternative?

1. Export a hook-based store via `create` (as opposed to the vanilla `createStore`).
2. This store exports an initialization function that populates the store with the passed data parameter.
3. The initialization function is passed the data to be injected and is called within a `useEffect` which runs only once.

For a more concrete example, consider the case of receiving server-side props from Next.js and needing to initialize a store with that data.

I'm assuming there's a downside to this hook-based approach that I'm missing, since all examples I've seen suggest context. Thank you in advance for your thoughts!

**Comment by dai-shi** [comment, marked as answer]
I think it's fine as long as the behavior with `useEffect` is okay: It re-renders twice initially. It's the downside.
